### DISCUSSION TITLE
I want to listen to the sound coming in on HDMI IN.

### DISCUSSION THREAD

**Discussion by ironiris** [Joshua-Riek/ubuntu-rockchip#199] [Q&A] [answered]
When I run the command below, I can see the video, but not hear the sound.
Video
`gst-launch-1.0 v4l2src device=/dev/video0 ! video/x-raw,format=NV12,width=3840,height=2160,framerate=60/1 ! autovideosink`
Audio
`gst-launch-1.0 alsasrc device=hw:4,0 ! audioconvert ! audioresample ! alsasink device=hw:0,0`

Is there a command that will allow me to hear the sound?
If I change all the numbers in the last hw:0,0 to hw:1,0, etc. with the command below, I can't hear the sound.

Just out of curiosity, I found that if I specify alsasrc as audiotestsrc instead of hdmiin, I can hear the test sound.
`gst-launch-1.0 alsasrc audiotestsrc ! audioconvert ! audioresample ! alsasink device=hw:0,0`

**Comment by Joshua-Riek** [comment]
I have been investigating sound issues. There is an issue with audio when using the usb-c display and an issue with the microphone. I have made some progress. However, my attempts to fix this bring back older problems, so I'm still looking for solutions. 

You can try setting `device=hw:0,0` to a different device for HDMI in:
https://en.wikibooks.org/wiki/Configuring_Sound_on_Linux/HW_Address#:~:text=The%20stock%20ALSA%20utility%20aplay%20can%20tell%20you,%5BATI%20HDMI%5D%20Subdevices%3A%201%2F1%20Subdevice%20%230%3A%20subdevice%20%230

**Comment by Joshua-Riek** [comment]
I'm looking at HDMI IN now. I found this helpful document from Rockchip.

[Rockchip_Developer_Guide_HDMI_RX_CN.pdf](https://github.com/Joshua-Riek/ubuntu-rockchip/files/11857386/Rockchip_Developer_Guide_HDMI_RX_CN.pdf)

**Reply by Joshua-Riek** [reply]
It might be a kernel driver problem. I see a lot of messages saying `rockchip-i2s-tdm fddf8000.i2s: ASoC: error at snd_soc_dai_hw_params on fddf8000.i2s: -22`, which relates to my below test for audio recording. 

```
ubuntu@ubuntu-desktop:~$ arecord -l
**** List of CAPTURE Hardware Devices ****
card 3: rockchipes8388 [rockchip,es8388], device 0: dailink-multicodecs ES8323.7-0011-0 [dailink-multicodecs ES8323.7-0011-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 4: rockchiphdmiin [rockchip-hdmiin], device 0: fddf8000.i2s-dummy_codec hdmiin-dc-0 [fddf8000.i2s-dummy_codec hdmiin-dc-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
ubuntu@ubuntu-desktop:~$ arecord -Dhw:4,0 -d 5 -f cd -r 44100 -c 2 -t wav audio.wav 
Recording WAVE 'audio.wav' : Signed 16 bit Little Endian, Rate 44100 Hz, Stereo
arecord: set_params:1416: Unable to install hw params:
ACCESS:  RW_INTERLEAVED
FORMAT:  S16_LE
SUBFORMAT:  STD
SAMPLE_BITS: 16
FRAME_BITS: 32
CHANNELS: 2
RATE: 44100
PERIOD_TIME: (125011 125012)
PERIOD_SIZE: 5513
PERIOD_BYTES: 22052
PERIODS: 4
BUFFER_TIME: (500045 500046)
BUFFER_SIZE: 22052
BUFFER_BYTES: 88208
TICK_TIME: 0
```

**Comment by Joshua-Riek** [comment]
I made more progress today. Somewhere between the rkr3.6 and rkr4 kernel, HDMI IN sound broke.

**Reply by ironiris** [reply]
It's something I feel all the time, but I really appreciate it.

**Reply by Joshua-Riek** [reply]
I asked around, and another developer might have a fix for this! I can test it out later today.

**Comment by Joshua-Riek** [comment]
Good news, I talked to another dev and they where able to come up with a fix for HDMI IN audio playback! I need to wait for the patch to be merged into the kernel. My current plan is to release a kernel update in 1-2 days with this fix.

**Reply by Joshua-Riek** [reply, marked as answer]
The new kernel has been uploaded. Please update your system with apt and then reboot once finished.

I used the below script to test the video and audio capture of HDMI input.
```bash
#!/bin/bash

trap 'onCtrlC' INT
function onCtrlC () {
	echo 'Ctrl+C is captured'
	killall gst-launch-1.0
	exit 0
}

device_id=$(v4l2-ctl --list-devices | grep -A1 hdmirx | grep -v hdmirx | awk -F ' ' '{print $NF}')
v4l2-ctl -d $device_id --set-dv-bt-timings query 2>&1 > /dev/null
width=$(v4l2-ctl -d $device_id --get-dv-timings | grep "Active width" |awk -F ' ' '{print $NF}')
heigh=$(v4l2-ctl -d $device_id --get-dv-timings | grep "Active heigh" |awk -F ' ' '{print $NF}')

es8388_card=$(aplay -l | grep "es8388" | cut -d ':' -f 1 | cut -d ' ' -f 2)
hdmi0_card=$(aplay -l | grep "hdmi0" | cut -d ':' -f 1 | cut -d ' ' -f 2)
hdmi1_card=$(aplay -l | grep "hdmi1" | cut -d ':' -f 1 | cut -d ' ' -f 2)
hdmiin_card=$(arecord -l | grep "hdmiin" | cut -d ":" -f 1 | cut -d ' ' -f 2)

DISPLAY=:0.0 gst-launch-1.0 v4l2src device=${device_id} io-mode=4 ! videoconvert \
	! video/x-raw,format=NV12,width=${width},height=${heigh} \
	! videoscale ! video/x-raw,width=1280,height=720 \
	! autovideosink sync=false 2>&1 > /dev/null &

gst-launch-1.0 alsasrc device=hw:${hdmiin_card},0 ! audioconvert ! audioresample ! queue \
	! tee name=t ! queue ! alsasink device="hw:${hdmi0_card},0" \
	t. ! queue ! alsasink device="hw:${hdmi1_card},0" &

while true
do
	sleep 10
done
```

**Reply by ironiris** [reply]
Perfect! Thank you.

**Reply by lvqier** [reply]
> Good news, I talked to another dev and they where able to come up with a fix for HDMI IN audio playback! I need to wait for the patch to be merged into the kernel. My current plan is to release a kernel update in 1-2 days with this fix.

Hi @Joshua-Riek I am struggling the same problem with rk-6.1-rkr3. Could you please guide me which patch might fixes? Appreciate for your reply!
